### PR TITLE
DT-1161: Collection entity page - display Group Members

### PR DESCRIPTION
**Pull request by YD-DLUHC**
**Description** 

This PR creates a new endpoint to take in a collection group/organisation-id and use them to retrieve the members of that group with their roles. 

The endpoint is set to filter only active users and those that are external.

### CODE REVIEW

**Review by hugh-emerson — CHANGES_REQUESTED** [review]

File: auth-service/src/main/kotlin/uk/gov/communities/delta/auth/config/LDAPConfig.kt [hunk 1]
```diff
@@ -35,6 +37,14 @@ data class LDAPConfig(
                 "ACCESS_GROUP_CONTAINER_DN",
                 "CN=datamart-delta,OU=Groups,OU=dluhctest,DC=dluhctest,DC=local"
             ),
+            groupContainerDn = Env.getRequiredOrDevFallback(
+                "ACCESS_GROUP_CONTAINER_DN",
```
[the remark this answers is not in the document]
Comment (YD-DLUHC) [reply auth-service/src/main/kotlin/uk/gov/communities/delta/auth/config/LDAPConfig.kt]: Good catch - Changed the name

File: auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt [hunk 2]
```diff
@@ -129,6 +131,47 @@ class EditAccessGroupsController(
...
+        missingFieldError?.let {
+            throw ApiError(HttpStatusCode.BadRequest, it.first, it.second)
+        }
+    }
+
+    private fun validateAccessGroupName(accessGroupName: String) {
+        val datamartDeltaPrefix = "datamart-delta-"
+        val validGroupNameRegex = Regex("^[a-z][a-z0-9\\-]+")
+
+        if (!validGroupNameRegex.matches(accessGroupName) || accessGroupName.startsWith(datamartDeltaPrefix)) {
```
[the remark this answers is not in the document]
Comment (YD-DLUHC) [reply auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt]: True - I've put in separate messages for each

File: auth-service/src/main/kotlin/uk/gov/communities/delta/auth/Routing.kt [hunk 3]
```diff
@@ -359,6 +359,9 @@ fun Route.bearerTokenRoutes(
             post("/access-groups/remove") {
                 editAccessGroupsController.removeUserFromAccessGroup(call)
             }
+            post("/access-groups/members") {
```
[the remark this answers is not in the document]
Comment (YD-DLUHC) [reply auth-service/src/main/kotlin/uk/gov/communities/delta/auth/Routing.kt]: Changed to GET and put the code in its own controller

File: auth-service/src/main/kotlin/uk/gov/communities/delta/auth/repositories/LdapRepository.kt [hunk 4]
```diff
@@ -35,6 +35,9 @@ class LdapRepository(
 
     private val logger = LoggerFactory.getLogger(javaClass)
     private val groupDnToCnRegex = Regex(ldapConfig.groupDnFormat.replace("%s", "([\\w-]+)"))
+    private val datamartDeltaPrefix = "datamart-delta-"
```
[the remark this answers is not in the document]
Comment (YD-DLUHC) [reply auth-service/src/main/kotlin/uk/gov/communities/delta/auth/repositories/LdapRepository.kt]: Changed to use existing prefix

File: auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt [hunk 5]
```diff
@@ -129,6 +131,47 @@ class EditAccessGroupsController(
...
+        }
+
+        missingFieldError?.let {
+            throw ApiError(HttpStatusCode.BadRequest, it.first, it.second)
+        }
+    }
+
+    private fun validateAccessGroupName(accessGroupName: String) {
+        val datamartDeltaPrefix = "datamart-delta-"
+        val validGroupNameRegex = Regex("^[a-z][a-z0-9\\-]+")
```
[the remark this answers is not in the document]
Comment (YD-DLUHC) [reply auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt]: Changed to use a public function as suggested

File: auth-service/src/test/kotlin/uk/gov/communities/delta/service/LDAPRepositoryTest.kt [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+import javax.naming.directory.*
+import javax.naming.ldap.InitialLdapContext
+
+class LdapRepositoryTest {
+
+    private val ldapConfig = mockk<LDAPConfig>(relaxed = true)
+    private val ctx = mockk<InitialLdapContext>()
+    private val ldapRepository = LdapRepository(ldapConfig, LdapRepository.ObjectGUIDMode.NEW_JAVA_UUID_STRING)
+
+//    @Test
```
Comment (hugh-emerson) [inline auth-service/src/test/kotlin/uk/gov/communities/delta/service/LDAPRepositoryTest.kt:16]: Why's this commented out?

Comment (YD-DLUHC) [reply auth-service/src/test/kotlin/uk/gov/communities/delta/service/LDAPRepositoryTest.kt:16]: I couldn't quite the test working but didn't want to spend too long behind it. Il remove it for now

Comment (keithkennedy) [reply auth-service/src/test/kotlin/uk/gov/communities/delta/service/LDAPRepositoryTest.kt]: Could we add a ticket to re-instate and fix this removed test, if it's useful?

Comment (keithkennedy) [reply auth-service/src/test/kotlin/uk/gov/communities/delta/service/LDAPRepositoryTest.kt]: Could we add a ticket to re-instate and fix this removed test, if it's useful?

**Review by keithkennedy — APPROVED** [review]

**Pull request merged**